### DISCUSSION TITLE
ConR analyse criterion B

### DISCUSSION THREAD

**Discussion by pvictor** [dreamRs/conrappli#8] [General]
J'ai extrait une partie des analyses à effectuer de l'application dans ce script : https://github.com/dreamRs/conrappli/blob/master/dev/ConR-analysis.R

Il y a qques problèmes avec les données enregistrées, je vais regarder plus en détail.

Est-ce que ce sont bien les analyses à effectuer dans l'application ?

Y-a-t-il beaucoup de changement par rapport à la version en développement de ConR ?

@gdauby @NicoTex

**Comment by gdauby** [comment]
Oui le package a assez bien évolué. Par exemple la fonction `IUCN.eval` n'existe plus.

Je vais mettre les exemples ci dessous fonction par fonction avec un commentaire pour les arguments pour lesquels il faudrait donner la possibilité à l'utilisateur de choisir entre différentes options/valeurs


## generation d'un jeu de donnees artificel
```
set.seed(3)
nbe_occ <- sample(seq(3, 100, 1), 20, replace = TRUE) ### 500 species with number of occurrences sampled between 3 and 100
ddlat <- lapply(nbe_occ, function(x) sample(seq(-10, 15, 0.1), x))
ddlon <- lapply(nbe_occ, function(x) sample(seq(7, 25, 0.1), x))
names(ddlon) <- names(ddlat) <- paste0("tax", seq(1,length(ddlat), 1))
ddlat <- lapply(seq_along(ddlat), function(i) data.frame(ddlat =  ddlat[[i]]))
ddlon <- lapply(seq_along(ddlon), function(i) data.frame(ddlon =  ddlon[[i]], taxa = names(ddlon)[i]))
test_data <- 
  data.frame(ddlat = do.call('rbind', ddlat), ddlon = do.call('rbind', ddlon)[,1], taxa = do.call('rbind', ddlon)[,2])
head(test_data)
dim(test_data)
```

```
p1 <- 
  EOO.computing(XY = test_data, 
                mode = "planar",  # input$mode_eoo a définir par utilisateur : 'spheroid' or 'planar'. By default 'spheroid'
                export_shp = TRUE)
```

`p1$spatial ` ## contient les polygones sous format sf, à utiliser par la suite si l'on veut 
`p1$results` ## les valeurs à garder car seront utilisées par la suite

**Comment by gdauby** [comment]
```
aoo_res <- AOO.computing(XY = test_data, 
              Cell_size_AOO = 2, # input$aoo_size min=0.1, max=50, value = 2, round=TRUE, step=1
              nbe.rep.rast.AOO = 10,  # input$rep_rast min = 0, max = 30, value = 10, round=TRUE, step=1
              export_shp = TRUE)
```

`aoo_res$AOO `## resultats à garder
`aoo_res$AOO_poly` ## polygones à garder si l'on veut cartographier les resultats, à voir selon les sorties que l'on pourra faire

**Comment by gdauby** [comment]
Pour info, je suis en train d'apporter des modifications relativement mineures à ConR, c'est pour cela que je n'ai pas encore publié le workflow complet:

Voici néanmoins ce à quoi devrait ressembler le workflow, sujet à modifications mineures donc.

```
## generation d'un jeu de donnees artificel pour reproduire/tester le workflow
set.seed(3)
nbe_occ <- sample(seq(1, 30, 1), 20, replace = TRUE) ### 500 species with number of occurrences sampled between 3 and 100
ddlat <- lapply(nbe_occ, function(x) sample(seq(-10, 15, 0.1), x))
ddlon <- lapply(nbe_occ, function(x) sample(seq(7, 25, 0.1), x))
names(ddlon) <- names(ddlat) <- paste0("tax", seq(1,length(ddlat), 1))
ddlat <- lapply(seq_along(ddlat), function(i) data.frame(ddlat =  ddlat[[i]]))
ddlon <- lapply(seq_along(ddlon), function(i) data.frame(ddlon =  ddlon[[i]], taxa = names(ddlon)[i]))
test_data <- 
  data.frame(ddlat = do.call('rbind', ddlat), ddlon = do.call('rbind', ddlon)[,1], taxa = do.call('rbind', ddlon)[,2])


eoo_res <-
  EOO.computing(XY = test_data, 
                mode = "planar",  # input$mode_eoo a définir par utilisateur : 'spheroid' or 'planar'. By default 'spheroid'
                export_shp = TRUE)

eoo_res$spatial
eoo_res$results

aoo_res <- AOO.computing(XY = test_data, 
              Cell_size_AOO = 2, # input$aoo_size min=0.1, max=50, value = 2, round=TRUE, step=1
              nbe.rep.rast.AOO = 0,  # input$rep_rast min = 0, max = 30, value = 10, round=TRUE, step=1
              export_shp = T)

locations <- locations.comp(XY = test_data, 
                            Cell_size_locations = 10 # input$locations_size min=0.1, max=50, value = 10, round=TRUE, step=1
                            )

categories <- 
  cat_criterion_b(EOO = eoo_res$results$eoo, AOO = aoo_res$AOO$aoo, locations = locations$locations$locations)

results_full <-
    data.frame(
      taxa = row.names(aoo_res$AOO),
      EOO = eoo_res$results$eoo,
      AOO = aoo_res$AOO$aoo,
      locations = locations$locations$locations,
      category = categories$ranks_B,
      cat_codes = categories$cats_code,
      issue_aoo = aoo_res$AOO$issue_aoo,
      issue_eoo = eoo_res$results$issue_eoo,
      issue_locations = locations$locations$issue_locations
    )

```
`results_full ` devrait être une sortie à afficher comme tableur après cette étape. Il serait bien aussi de pouvoir télécharger cette table à cette étape.

**Reply by pvictor** [reply]
Bonjour Gilles,

J'obtiens une erreur dans la fonction `locations.comp()` en utilisant les données du CSV attaché (il s'agit des données homemade nettoyées après les étapes de validation)

```r
test_data <- readr::read_csv("exemple_homemade_cleaned_data.csv")
locations <- locations.comp(
  XY = test_data,
  Cell_size_locations = 10
)
# |=================================================================================================================================================================================| 100%
# Error in { : 
#     task 21 failed - "arguments imply differing number of rows: 0, 1"
```

[exemple_homemade_cleaned_data.csv](https://github.com/dreamRs/conrappli/files/8621706/exemple_homemade_cleaned_data.csv)

**Comment by gdauby** [comment]
Bonjour,
j'y ai apporté des modifications fin de semaine dernière, je viens de pushé la nouvelle version 4.0.2. Ca devrait fonctionner avec cette version.

Il pourrait y avoir des petites modifications à faire dans le workflow que j'ai posté la semaine dernière, je vais vérifier.

**Comment by gdauby** [comment]
Bonjour,
j'ai mis à jour le workflow ci dessus, en lien avec la nouvelle version.

**Comment by gdauby** [comment]
@pvictor voici une mise à jour du workflow pour l'analyse ConR du module criterion B

Jeu de données pour tester au cas ou
```
set.seed(3)
nbe_occ <- sample(seq(1, 15, 1), 10, replace = TRUE) ### 500 species with number of occurrences sampled between 3 and 100
ddlat <- lapply(nbe_occ, function(x) sample(seq(-10, 15, 0.1), x))
ddlon <- lapply(nbe_occ, function(x) sample(seq(7, 25, 0.1), x))
names(ddlon) <- names(ddlat) <- paste0("tax", seq(1,length(ddlat), 1))
ddlat <- lapply(seq_along(ddlat), function(i) data.frame(ddlat =  ddlat[[i]]))
ddlon <- lapply(seq_along(ddlon), function(i) data.frame(ddlon =  ddlon[[i]], taxa = names(ddlon)[i]))
test_data <- 
  data.frame(ddlat = do.call('rbind', ddlat), ddlon = do.call('rbind', ddlon)[,1], taxa = do.call('rbind', ddlon)[,2])
```


```
eoo_res <-
  EOO.computing(XY = test_data, 
                mode = "planar",  # input$mode_eoo a définir par utilisateur : 'spheroid' or 'planar'. By default 'spheroid'
                export_shp = TRUE)

eoo_res$spatial ### data a garder
eoo_res$results ### data a garder

aoo_res <- AOO.computing(XY = test_data, 
              Cell_size_AOO = 2, # input$aoo_size min=0.1, max=50, value = 2, round=TRUE, step=1
              nbe.rep.rast.AOO = 0,  # input$rep_rast min = 0, max = 30, value = 10, round=TRUE, step=1
              export_shp = T)
```

Les fonctions `extract_overlap_shp `et `collect_shp `sont donnés au point #10 

Garder cette table entière et afficher la table (les 4 colonne sélectionnés)
```
check_overlap <- extract_overlap_shp(XY = test_data)

check_overlap$shp_tables %>% 
  dplyr::select(id, table_name, type, description, reference) 

```

Si au moins une donnée spatial se superspose avec les données `any(check_overlap$res)` est `TRUE`, alors l'utilisateur peut sélectionner  maximum 5 jeux de données spatiaux minimum 0 (l'utilisateur peut décider de ne choisir aucun jeu de données). Un jeu de données, ne peut pas être choisi deux fois.

Je ne sais pas quel est la meilleure fonction shiny pour ça
peut etre `selectizeInput` ?
On peut aussi imaginer un système comme celui de la sélection des variables.

Une fois la sélection faite, il faut aller collecter ces shapefiles sur le serveur avec la fonction suivante (ici pour tous les shapefiles qui se superposent).

```
if (any(check_overlap$shp_tables$overlap)) {
  all_shp <-
    collect_shp(table_names = check_overlap$shp_tables[which(check_overlap$shp_tables$overlap),], 
                XY_sf = check_overlap$XY_sf)
} else {
  all_shp <-
    NULL
}
```

L'ordre dans lequel les données spatiales sont choisis a une importance : il faut re arranger la liste `all_shp ` selon la sélection faite par l'utilisateur.

Garder cette liste, même si aucune données spatiales n'a été choisie car elle sera utilisée dans la deuxième partie des analyses (Criterion A)
`all_shp `


Finalement, ces shapefiles sont utilisés dans le calcule des location avec le paramètre `threat_list `.
Il y a un paramètre à rajouter dans les `input` pour `method_polygons `.
```
locations <- locations.comp(XY = test_data, 
                            Cell_size_locations = 10,  # input$locations_size min=0.1, max=50, value = 10, round=TRUE, step=1
                            threat_list =  all_shp, 
                            method_polygons = "no_more_than_one"  # input$method_polygons a définir par utilisateur : 'spheroid' or 'planar'. By default 'spheroid'
                            )
```

`locations$locations_poly`  Garder cette table

```
categories <- 
  cat_criterion_b(EOO = eoo_res$results$eoo, 
                  AOO = aoo_res$AOO$aoo, 
                  locations = locations$locations$locations)

results_full <-
    data.frame(
      taxa = row.names(aoo_res$AOO),
      EOO = eoo_res$results$eoo,
      AOO = aoo_res$AOO$aoo,
      locations = locations$locations$locations,
      category = categories$ranks_B,
      cat_codes = categories$cats_code,
      issue_aoo = aoo_res$AOO$issue_aoo,
      issue_eoo = eoo_res$results$issue_eoo,
      issue_locations = locations$locations$issue_locations
    )

``` 

Objets à garder pour l'exportation de résultats en `rmarkdown`
```
aoo_res$AOO_poly
eoo_res$spatial
locations$locations_poly
```

**Reply by pvictor** [reply]
Merci Gilles, je regarde cela
